app/ts/background/metadataUtils.ts

```diff
@@ -200,15 +200,35 @@ export const extractTokenEvents = (events: readonly EnrichedEthereumEventWithMet
 
 export async function retrieveEnsNodeHashes(ethereumClientService: EthereumClientService, events: EnrichedEthereumEvents) {
 	const hashes = events.map((event) => event.type === 'ENSAddrChanged' || event.type === 'ENSAddressChanged' ? event.logInformation.node : undefined).filter((maybeNodeHash): maybeNodeHash is bigint => maybeNodeHash !== undefined)
-	const deduplicated = Array.from(new Set(hashes))
-	return Promise.all(deduplicated.map((hash) => getAndCacheEnsNodeHash(ethereumClientService, hash)))
+	const deduplicatedHashes = Array.from(new Set(hashes))
+	return await Promise.all(deduplicatedHashes.map((hash) => getAndCacheEnsNodeHash(ethereumClientService, hash)))
 }
 
-export const getAndCacheEnsNodeHash = async (ethereumClientService: EthereumClientService, nameHash: EthereumBytes32) => {
+export async function retrieveEnsLabelHashes(events: EnrichedEthereumEvents) {
+	const labelHashesToRetrieve = events.map((event) => event.type === 'ENSNameRenewed' || event.type === 'ENSRegistrarNameRenewed' ? event.logInformation.labelHash : undefined).filter((labelHash): labelHash is bigint => labelHash !== undefined)
+	const newLabels = events.map((event) => event.type === 'ENSRegistrarNameRenewed' ? event.logInformation.name : undefined).filter((label): label is string => label !== undefined)
+	
+	// update the maping if we have new labels
+	const deduplicatedLabels = Array.from(new Set(newLabels))
+	await Promise.all(deduplicatedLabels.map(async (label) => await addEnsLabelHash(label)))
+	
+	// return the label hashes that we have now available
+	const currentLabelHashes = await getEnsLabelHashes()
+	return Array.from(new Set(labelHashesToRetrieve)).map((labelHash) => {
+		const found = currentLabelHashes.find((entry) => entry.labelHash === labelHash)
+		return { labelHash, label: found?.label }
+	})
```

Comment:
Looping the event log should be performant in getting info from it, also by initiating the reference as a `Set`, would by nature deduplicate the result.

```suggestion
	const labelHashesToRetrieve = new Set<bigint>()
	const newLabels = new Set<string>()

	for (const event of events) {
		if (event.type === 'ENSNameRenewed' || event.type === 'ENSRegistrarNameRenewed') {
			labelHashesToRetrieve.add(event.logInformation.labelHash)
		}
		if (event.type === 'ENSRegistrarNameRenewed') {
			newLabels.add(event.logInformation.name)
		}
	}

	await Promise.all([...newLabels].map(async (label) => await addEnsLabelHash(label)))

	const currentLabelHashes = await getEnsLabelHashes()
	return [...labelHashesToRetrieve].map((labelHash) => {
		const found = currentLabelHashes.find((entry) => entry.labelHash === labelHash)
		return { labelHash, label: found?.label }
	})
```

Reply:
I don't think we should program these things with efficiency in mind as these are quite fast operations. I think adding to set one by one is anyway slower than sorting it once.